src/main/kotlin/racing/Car.kt

```diff
@@ -1,11 +1,22 @@
 package racing
 
-class Car {
+class Car(val name: String) {
     private var turn = 0
 
     val turnOfPosition = arrayListOf(1)
 
-    fun go() {
-        turnOfPosition.add(turnOfPosition[turn++] + if ((0..9).random() >= 4) 1 else 0)
+    fun goIfAboveRequireGoValue(): Int {
+        return if ((0..9).random() >= REQUIRE_GO_VALUE) GO else STAY
+    }
+
+    fun go(goIfAboveRequireGoValue: () -> Int) {
```

Comment:
함수타입 파라메터를 활용해 보셨군요! 👍 

그런데 이 파라메터 명이 너무 구체적이어서 이 람다의 내용은 반드시 `goIfAboveRequireGoValue` 함수만 써야할것 같은 느낌이 들어요. 한편 `Car` 가 이동하는 조건을 노출하게 되어서 `Car` 객체에게 `go` 메시지를 보내는 클라이언트 코드는 불필요한 정보를 알게 되어 결국 캡슐화가 지켜지지 않았어요.

```suggestion
    fun go(distance: () -> Int) {
```

이렇게 바꾼다면 `distance: () -> Int` 함수는 항상 1만 나오는, 한상 0만 나오는, 혹은 랜덤하게 1 또는 0이 나오거나 2 이상이 나올수도 있는 확장성을 얻게 되어요. 다른말로는 OCP 라고도 합니다.

> 다만 go 함수에 () -> Int 이용해서 파라미터를 넘기고 났을 때 테스트코드를 어떻게 작성해야하는지는 감이 잘 안오네요ㅜㅜ.. 리뷰에 힌트 주시면 보고 확인해서 또 작성해보도록 하겠습니다..!

`Car` 객체에게 go 메시지를 보내면 `Car` 객체의 상태가 바뀌게 될텐데요. 예상되는 경계값들을 통해 기대한대로 자동차가 움직이는지 테스트 해보면 어떨까요?

Reply:
Car 객체에게 go 메시지를 보낸다는 것이 어떤 의미일까요??

> Car 가 이동하는 조건을 노출하게 되어서 Car 객체에게 go 메시지를 보내는 클라이언트 코드는 불필요한 정보를 알게 되어 결국 캡슐화가 지켜지지 않았어요.
이 부분이 이해가 되지 않아서 질문 드립니다!

Reply:
> Car 객체에게 go 메시지를 보낸다는 것이 어떤 의미일까요??

```kotlin
car.go { car.goIfAboveRequireGoValue() }
```

위 코드가 `CarRacing` 객체가 `car` 객체에게 이동하라는 메시지를 모내는 부분입니다. 위 코드는 아래처럼 바꿀수도 있습니다.
```kotlin
car.go(car::goIfAboveRequireGoValue)
```
메시지와 함께 `car:goIfAboveRequireGoValue` 를 포낸다는 의미는 `CarRacing` 객체가 `car` 객체가 메시지를 보내는 시점에 어떻게 움직어야 하는지 결정하고 있다는 의미입니다. 객체지향 관점에서는 굳이 알아야 할 정보가 아니라는 의미에요. `car` 객체에게 `go` 라는 메시지만 보낼 뿐 이동여부는 car 객체가 스스로 판단해서 결정해야 한다고 생각합니다. 만약 `Car` 객체의 이동방식이 바뀐다면 반드시 `go` 함수를 호출하는 `moveCars` 도 바뀌어야 할 것입니다. 변경에 닫혀있지 않으므로 OCP 를 지키지 못하고 있습니다. 

그런데 글을 작성하다보니 위와 같은 주장이 무조건 옳은건 아니라는 생각도 듭니다. `CarRacing` 객체가 `Car` 객체의 행동방식을 결정할 수도 있지요. ㅎㅎ 두 객체를 강하게 의존시키려는 의도가 있다면, `Car` 객체가 어떻게 움직이는지 `CarRacing` 에게 중요한 정보라면 의도적으로 노출시키는것이 더 좋을수도 있을것 같습니다. 

아래 부분은 아실수도 있지만 혹시 말씀드린 부분에 더 자세한 설명이 필요할지 몰라서 덧붙입니다.

> `Car` 객체의 `go`함수를 호출하는것은 `Car` 객체를 의존하는 객체(`CarRacing`) 가 `Car` 객체에게 이동(`go`)하라는 메시지를 모낸다는 것입니다. 여기서 **명령** 이 아니라 **메시지**라고 표현한데에 주목해 주세요. 명령은 대상이 시키는대로 항상 따라야 하는 수동적인 존재로 보는 것입니다. 그러나 메시지는 대상이 메시지를 따를지 따르지 않을지 스스로 결정할 수 있음을 인정한다는 의미입니다. 이와 관련하여 더 자세하게 설명된 책은 [일명 토끼책](http://www.yes24.com/Product/Goods/18249021)을 보시면 잘 설명되어 있습니다.

Reply:
완전 이해 완료하였습니다!!! 설명 감사드립니다!! 
두 객체의 의존성에 대해 다시 한 번 생각해보도록 할게요!